Refactor codebase to be a monorepo project

ferreiro opened the pull request:
In this PR I'm creating a react app for the project, and in order to be able to have independent `client` and `server` I'm updating the project to use a mono-repo approach

In order to connect react app with the server I already have in place I'm following [this tutorial](https://medium.freecodecamp.org/how-to-make-create-react-app-work-with-a-node-backend-api-7c5c48acb1b0) which uses the same stack as this app (heroku, node...).

### Known issues
## Issue 1: When `$ yarn start` inside `/web-app`, it was trying to resolve dependencies by looking inside the root node_modules.

I got some troubles that I fixed, but I've doubts about them. Basically, there was dependency conflicts between `web-app` and the project root dependencies. [Any time I run `yarn start` from `web-app` it threw an error. I think it was trying to get dependencies from the root `node_modules` instead of `web-app/node_modules`](https://gist.github.com/ferreiro/a796a5031fcd6e33b148e7ae1975e733) => That's weird!!

I solved that by removing the following dependencies from the project (standard, webpack, validator). But I wanna know why I was getting that error and why dependencies are resolved in that way

## Issue 2: Proxy from `web-app/package.json` doesn't work, and fetch is trying to call the react app url instead of api.

[DEV mode] I'm running the API/Server in port 3000, and I'm running the react app in port 5000. In order to locally don't need to define `localhost:3000/api/v1`,[ I added a proxy](https://facebook.github.io/create-react-app/docs/proxying-api-requests-in-development). This is supposed to make any `fetch()` request from the react app (I added a dummy fetch in App.js) to call the localhost:3000 instead of the localhost:5000. I searched online but haven't found a solution.

ferreiro commented:
This CR is the first step to finish: https://github.com/ferreiro/website/pull/40

kwelch commented on package.json:
```diff
@@ -22,20 +22,21 @@
+    "lint": "",
+    "client": "cd web-app && PORT=5000 yarn start",
+    "server": "DEBUG=ferreiro:* nodemon --exec babel-node ./bin/www",
+    "dev": "concurrently --kill-others-on-fail \"yarn server\"  \"yarn build\"  \"yarn client\" \"yarn watch\"",
```
do you need to `gulp build` & `gulp watch` in the same command. My first though would be that `watch` is running `build` continuously.

kwelch commented on README.md:
```diff
@@ -2,6 +2,36 @@
 
+## Env variables
+
+We have all the configuration for beta and prod inside `./env/index.js`, there we load 2 files, the production or the development.
```
I have used `.env` before but never as a folder. I typically do it as `.env.development` and `.env.production`. Not saying this is wrong, just a different way. I using folders to contain similar concerns.

ferreiro replied:
Ops. Miss this message. THat's an interesting point. I followed this pattern from a book I read about Node and for me it looked like a nice way to be explicit about prod and dev, and provide default values.

[In heroku they propose to use `.env` files](https://devcenter.heroku.com/articles/node-best-practices#be-environmentally-aware)

I'm happy to revisit the way we handle env variables for this project. [So I've just created a ticket for it so we can revisit it later on.](https://github.com/ferreiro/website/issues/45)

kwelch replied:
That is a great plan.

kwelch commented on web-app/src/App.js:
```diff
@@ -0,0 +1,38 @@
+  componentDidMount = () => {
+      // This code should call: http://localhost:3000/api/status
+      // and return {"response":"Everything is working correctly"}
+      fetch('/api/status')
```
Why could you not just call over to localhost:5000? did you get a cors error?

ferreiro replied:
Yeap! I got a cors error. Also, the problem of using directly dev endpoints is that we'll need to change that once deployed (or implement some logic like: if DEVELOPMENT => localhost:3000, otherwise "ferreiro.me/". I think that's the purpose of the proxy, but not working)

```Access to fetch at 'http://localhost:3000/api/status' from origin 'http://localhost:5000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.```

kwelch commented:
# Issue 1
We may want to convert this to a yarn workspace to avoid the dependency issue. Typically module resolution would not go up a directory unless it was not found in the closest node_modules. So either it was not listed which seems unlikely, or it was not passing `process.cwd` the way you expected.

ferreiro commented:
> # Issue 1
> We may want to convert this to a yarn workspace to avoid the dependency issue. Typically module resolution would not go up a directory unless it was not found in the closest node_modules. So either it was not listed which seems unlikely, or it was not passing `process.cwd` the way you expected.

@kwelch OK! I'm updating this to use => https://yarnpkg.com/lang/en/docs/workspaces/#toc-how-to-use-it

ferreiro commented:
In the lasts commits I've refactored the logic to be monorepo!!!! 

Question: Do we want to have in the repo the `yarn.lock` and `packackage.lock`? Is there any good practice around this?

ferreiro commented:
Yay! Found in the last bunch of commits how to create a local package, and add it as a dependency to another package.

1. I created `design_system` package
2. Then in @ferreiro/client I run `yarn add /Users/ferreiro/Documents/website/packages/design_system`

Voila! Now, in order to use the local copy of the package I run `yarn link` inside the design_system package, and then inside `ferreiro-client` I run yarn link "@ferreiro/design-system" which is the name of the package.

ferreiro commented on package.json:
```diff
@@ -1,102 +1,21 @@
+        "node": "8.10.0"
+    },
+    "scripts": {
+        "client": "cd packages/ferreiro-client && PORT=5000 yarn start",
```
oh nice!

ferreiro commented on packages/ferreiro-client/package.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "@ferreiro/client",
```
ohh! That's true. I included to follow how babel and other open source projects use them. They always include the project name there. I also think that we can maybe release those packages to the npm (not sure if it's very helpful, but it's nice to work with npm workflow)

kwelch replied:
Yea, working on a release pipeline is great and will just give you more practice if you are ever looking to do that for a different project.

ferreiro commented on packages/ferreiro-client/package.json:
```diff
@@ -0,0 +1,30 @@
+  "version": "0.0.0",
+  "private": false,
+  "dependencies": {
+    "@ferreiro/design-system": "/Users/ferreiro/Documents/website/packages/design-system",
```
oh! Good point. I also don't like to include the local path. I played with `yarn link` too and it's super cool to be able to use a local copy of an external dependecy. 

That's said, since the `design_system` package is not yet published, what should I put in the value for `@ferreiro/design-system:`? I was confused about it.

kwelch replied:
Use the value from its `package.json`, yarn will handle the rest.

ferreiro commented on packages/ferreiro-client/src/App.js:
```diff
@@ -0,0 +1,45 @@
+    console.log('fetching...')
+      // This code should call: http://localhost:3000/api/status
+      // and return {"response":"Everything is working correctly"}
+      fetch('/api/status')
```
Nop! I think one of the problems is that concurrently launches both proceses and if the react client is initialized before the server, then the proxy is not binded. I saw in the react doc they suggest to use https://facebook.github.io/create-react-app/docs/proxying-api-requests-in-development#configuring-the-proxy-manually

I tried to use that manual proxy and put a delay in the process to wait till the api is mounted, but couldn't make it work.

Do you have some ideas? I'll try to play with it tomorrow as well, but it looks now is a bit of a blocker.

ferreiro commented on packages/design-system/package.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "private": true,
+  "name": "@ferreiro/design-system",
```
I also like that. Maybe my main question here is: do we want to create a "generic" design system that anybody in the world can use in their projects or do we only link them with this project?

I'm emvisioning kind of a design system that other devs can take advantage of. I saw `jds` is free in NPM, but not sure if it's the best option. What are your thoughts on this?

kwelch replied:
I don't think it is going to be possible to have a generic design system that covers everything. For now I would focus on your case. Then re-evaluate once you have it working.

ferreiro replied:
agree! Let's move forward with `@ferreiro/design-system` and later we can decide if it makes sense to release using `jds`

ferreiro commented on package.json:
```diff
@@ -1,102 +1,21 @@
+    "workspaces": [
+        "packages/*"
+    ],
+    "devDependencies": {}
```
Good point! I added concurrently to devDependencies in @ferreiro/server. Forgot to add it here.

Side question: Last time I checked, heroku doesn't care about devDependencies and it only gets the normal dependencies. So I think I should add concurrently here as well.

Side question 2: If concurrently doesn't appear as a dependency in the root package, is it because is getting the dependency that `ferreiro-server/package.json` instanciates? I don't think I have a local copy.

kwelch replied:
Yes, part of `yarn workspaces` is that it pull deps from the inner packages to the top level. Then we it has conflict it put specific version in the `node_modules` relative to that workspace.

ferreiro replied:
cool. Is there a way to run a script that checks that all the required dependencies for each of the project packages are installed? Otherwise, it looks we can have cases where yarn workspaces installs the dependency on another package and doesn't complain, right?

ferreiro commented on packages/design-system/README.md:
````diff
@@ -0,0 +1,19 @@
+# Naming
+## Requirements
+
+- Easy to call ```import {Button} from "@jds";```
````
true! I'd opt for `jds` in this case. I think is simpler and easier to call from the code. I like that version instead of typing `@ferreiro/jds` any time.

ferreiro commented:
@kwelch Thanks for the review. Answering your question below:

- How does this feel so far? Super cool! I loved a lot to work on this, which is not directly related with building new features but its SO IMPORTANT. That feeling where you told me about the `monorepo` was AMAZING. I'm really excited to also remove gulp and use webpack and set all the tests and so on.

- Any areas in particular that you feel like you are hung up? The one with the `proxy`. It's not working now. Also, I still need to understand better how the design system will interact with the app, and have some doubts about the comment about how I link the design system locally.

My main open question now is how are we going to integrate the react app with the current website. (SSR is another story :P)

- What do you see as your next steps? I wanna see this landed! hehe. Apart from that, I'd say write a blog post about creating a monorepo react app (I drafted yesterday), and then starting to create the first design system components (the card as we said and the small items that compose that card). In  fact, I started working with the card during the weekend too.

I still don't know if we wanna unit tests all the design system stuff, probably it would be great to do. While can slow me down a bit, it may be great if we want to create a design system that other users can grab.

kwelch commented:
>I still don't know if we wanna unit tests all the design system stuff, probably it would be great to do. While that can "slow me down" a bit, it may be great if we want to create a design system that other users can use and is well tested. Maybe is a good opportunity to start doing TDD?

I agree I think working on the design system as TDD is a good approach. Start out by putting up a diff of just tests for something like the card or button. The best way to do this is to envision how you want to use it and what it would do and write tests that match that. Then we can code it to work the way you hope to use it.

ferreiro commented:
<Discussed offline> In order to merge this CR, I'll need to make sure that we're able to deploy the website in Heroku. 

Note: We'll not work on the proxy right now. Probably will need in the future, if we decide not to have 2 separate servers running or... we move with a more complex solution like Dockerize the project or using Kubernetes

ferreiro commented:
@kwelch just updated the `package.json` script to be able to deploy the monorepo app into heroku. I created also a Procfile, since the instance I was using is aroung 512MB and I limited the size for the garbage collector following Heroku good practises: https://ferreiro-monorepo.herokuapp.com/

NB: I'm not initializing the React client server. I think in order to expose apps in different ports, we'll need to follow some of the ideas listed here like [using NGNX](https://www.quora.com/How-do-I-host-a-Node-js-app-that-listens-on-multiple-ports/answer/Baraka-Andrew) or [something to manage processes](https://medium.com/@nadayar/heroku-fu-multiple-servers-on-one-dyno-6fc68d57b373)

kwelch commented on packages/ferreiro-server/package.json:
```diff
@@ -96,7 +97,6 @@
     "validator": "^9.0.0"
   },
   "devDependencies": {
-    "concurrently": "^4.1.0",
```
I think this should still be a devDep, since it is not used within the actual app, but rather in the build process.

kwelch commented on package.json:
```diff
@@ -12,10 +12,13 @@
-    "devDependencies": {}
+    "devDependencies": {},
+    "dependencies": {
+        "concurrently": "^4.1.0"
```
This should be a devDep since it is not used in the running code.

ferreiro commented:
@kwelch Hey! Excited about this new revision. I think we're really close to launch to prod!

After some tweaks and updates, the webpage is up!!! https://ferreiro-monorepo.herokuapp.com/

Main changes:
- Removed all the react files from `ferreiro-client`.
- Analyzed dependencies and moved correctly into `devDependencies`.
- Removed unused dependencies. I took this opportunity to run `npm-check` and remove all the non-relevant packages from @ferreiro-server.

Some other changes in order to make heroku works:
- Set yarn version into the root `package.json` ([needed by heroku](https://devcenter.heroku.com/articles/nodejs-support#specifying-a-yarn-version)).
- Created `heroku-postbuild` and installing all the static assets there.
- Removed `yarn install` from `start` script, since Heroku already installs all the dependencies when found a yarn.lock, so that was failing the boot up of the server since 

Also, I needed to move `babel-cli, babel-core, babel-loader, babel-preset-env, babel-present-es2015` (https://medium.com/developer-circles-lusaka/how-to-setup-express-js-server-with-nodemon-and-babel-c3a17218c282) to `dependencies`, since once I run `yarn server` Heroku doesn't have the devDependencies (they delete all those dependencies once build).

kwelch commented on package.json:
```diff
@@ -1,102 +1,25 @@
+        "buildAssets": "yarn workspace @ferreiro/server build-assets",
+        "dev": "yarn install && yarn buildAssets && concurrently --kill-others-on-fail \"yarn devServer\"",
+        "start": "yarn server",
+        "heroku-postbuild": "npm run buildAssets"
```
running yarn in a bunch of places then npm seems odd.

ferreiro replied:
Done! Good catch :)

kwelch commented:
>Also, I needed to move babel-cli, babel-core, babel-loader, babel-preset-env, babel-present-es2015 (medium.com/developer-circles-lusaka/how-to-setup-express-js-server-with-nodemon-and-babel-c3a17218c282) to dependencies, since once I run yarn server Heroku doesn't have the devDependencies (they delete all those dependencies once build).

We could get around this but it would involve compiling the server before deploying. This is probably not necessary extra piece. Adding babel as a dep is not that bad.

ferreiro commented:
Merging this now! Thanks @kwelch for following up on this.

I leave the design system since is gonna be part of the next steps, so we can have that as the starting point.

ferreiro commented:
fixes #41

ferreiro commented:
Moving here the issues that I posted before in the description, since they are not relevant there anymore:


### Known issues
## Issue 1: When `$ yarn start` inside `/web-app`, it was trying to resolve dependencies by looking inside the root node_modules.

I got some troubles that I fixed, but I've doubts about them. Basically, there was dependency conflicts between `web-app` and the project root dependencies. [Any time I run `yarn start` from `web-app` it threw an error. I think it was trying to get dependencies from the root `node_modules` instead of `web-app/node_modules`](https://gist.github.com/ferreiro/a796a5031fcd6e33b148e7ae1975e733) => That's weird!!

I solved that by removing the following dependencies from the project (standard, webpack, validator). But I wanna know why I was getting that error and why dependencies are resolved in that way

## Issue 2: Proxy from `web-app/package.json` doesn't work, and fetch is trying to call the react app url instead of api.

[DEV mode] I'm running the API/Server in port 3000, and I'm running the react app in port 5000. In order to locally don't need to define `localhost:3000/api/v1`,[ I added a proxy](https://facebook.github.io/create-react-app/docs/proxying-api-requests-in-development). This is supposed to make any `fetch()` request from the react app (I added a dummy fetch in App.js) to call the localhost:3000 instead of the localhost:5000. I searched online but haven't found a solution.
